janus_server/src/binary_utils/job_driver.rs

```diff
@@ -0,0 +1,284 @@
+//! Discovery and driving of jobs scheduled elsewhere.
+
+use crate::{
+    datastore::{self, Datastore},
+    task::VdafInstance,
+};
+use janus::{
+    message::{Duration, TaskId, Time},
+    time::Clock,
+};
+use std::{fmt::Debug, future::Future, sync::Arc};
+use tokio::{sync::Semaphore, task, time};
+use tracing::{debug, error, info, info_span, Instrument};
+
+/// Periodically seeks incomplete jobs in the datastore and drives them concurrently.
+pub struct JobDriver<C: Clock, Acquirer, Stepper> {
+    /// Datastore where incomplete jobs are discovered and which is provided to the job driver.
+    datastore: Arc<Datastore<C>>,
+    /// Clock used to determine when to schedule jobs.
+    clock: C,
+
+    // Configuration values.
+    /// Minimum delay between datastore job discovery passes.
+    min_job_discovery_delay: Duration,
+    /// Maximum delay between datastore job discovery passes.
+    max_job_discovery_delay: Duration,
+    /// How many jobs to step at the same time in this process.
+    max_concurrent_job_workers: usize,
+    /// How long a lease on a job lasts.
+    worker_lease_duration: Duration,
+    /// Allowable clock skew between datastore and job driver, used when determining if a lease has
+    /// expired.
+    worker_lease_clock_skew_allowance: Duration,
+
+    // Callbacks.
+    /// Finds incomplete jobs in the datastore and acquires a lease on them.
+    incomplete_job_acquirer: Acquirer,
+    /// Steps an incomplete job.
+    job_stepper: Stepper,
+}
+
+impl<C, E, Acquirer, AcquirerFut, Stepper, StepperFut, JobId> JobDriver<C, Acquirer, Stepper>
+where
+    C: Clock,
+    E: Debug + Send + Sync + 'static,
+    Acquirer: Fn(Arc<Datastore<C>>, Duration, usize) -> AcquirerFut + Send + Sync + 'static,
+    AcquirerFut:
+        Future<Output = Result<Vec<(TaskId, VdafInstance, JobId, Time)>, datastore::Error>> + Send,
```

Comment:
I'm not sure this is the right abstraction for acquired jobs: I think we want something like `Vec<(UserDefinedStructure, Time)>`, i.e. replace all of `TaskId, VdafInstance, JobId` with a user-defined structure that ends up being passed to `Stepper`. I say this for a couple of reasons:

* I'm not sure `VdafInstance` is generally-applicable. The collect job driver is going to be making `AggregateShareReqs` as it can & storing the results in the datastore to eventually be included in a `CollectResp`, right? If so, I don't think there's anything that would require knowing the VDAF type. (the existing agg job driver retrieves the VDAF type here so it can do the generic-function match early, to drop one tx and/or some very awkward code in the stepper logic)
* I am not sure that every kind of job will be identified by a `TaskId` and one other kind of `JobId` -- eg we might have `TaskId, JobIdA, JobIdB`. While this could be worked around in the existing framework by making `JobId = (JobIdA, JobIdB)`, this is more awkward than just having the caller use a structure of their own. In the other direction, if we have a task-level driver, we wouldn't need `JobId` at all; using a user-defined structure in this case avoids the awkwardness of `JobId = ()`.